src/main/java/org/jenkinsci/plugins/workflow/steps/scm/SCMStep.java

```diff
@@ -138,6 +139,10 @@ public final void checkout(Run<?,?> run, FilePath workspace, TaskListener listen
                 } catch (InterruptedIOException e) {
                     throw e;
```

Comment:
I'm curious on what's the purpose of this. Perhaps it can be removed now.

/cc @dwnusbaum

Reply:
There are no more references in the code about it:

![image](https://github.com/jenkinsci/workflow-scm-step-plugin/assets/29582865/0d857c54-3b3d-44ea-9369-6a5481d97fd5)

Perhaps it was meant to do what this PR actually does.

Reply:
Out of curiosity, the checkout to get Jenkinsfile step does the same thing:

https://github.com/jenkinsci/workflow-cps-plugin/pull/147/files#diff-751ef0b8272fefc4b69f6302b381e99ddc17417176a159adaf9e98e5d83566d4R144

And it is probably affected by the same problem.

Reply:
> And it is probably affected by the same problem.

I just confirmed here, it is. I'll craft a similar fix there.

Reply:
You would have to look through all implementations of `SCM` in other plugins and see if their `checkout` methods ever throw `InterruptedIOException` (keeping in mind that they might not throw it directly, but might use IO-related JDK APIs that in turn throw `InterruptedIOException`) to really know if it is unused or not. 

Really, looking at https://github.com/jenkinsci/workflow-cps-plugin/pull/147, I think the better fix would be to change the `catch (Exception e)` block here to `catch (IOException e)`, so that instances of `InterruptedException` (which `FlowInterruptedException` includes) always bypass the retry logic, and then add a catch block for `AbortException`. In other words, use the exact same try/catch logic as in https://github.com/jenkinsci/workflow-cps-plugin/pull/147.

Reply:
Oh, but then yes https://github.com/jenkinsci/workflow-cps-global-lib-plugin/pull/63 made `workflow-cps` have the same bad behavior. And then compare the third implementation over in https://github.com/jenkinsci/pipeline-groovy-lib-plugin/blob/83f864f66d8f827c442913c4e3f7000c3a52b5fb/src/main/java/org/jenkinsci/plugins/workflow/libs/SCMBasedRetriever.java#L218. Probably we need to create a helper method in `workflow-api` for this retry logic that handles all of the cases correctly.

Reply:
> use the exact same try/catch logic as in https://github.com/jenkinsci/workflow-cps-plugin/pull/147.

But I just confirmed that one doesn't work either. It's affected by the same issue.

Reply:
IMO the logic seems like it should be like this: instances of `InterruptedIOException` and `InterruptedException` should bypass retries, except we need to check `FlowInterruptedException`, which is a subclass of `InterruptedException`, specially, and only skip retries if `isActualInterruption()`.

Then, for any other `Exception`, we would retry, handling `AbortException` specially as needed to avoid printing stack traces for them.

Reply:
> Probably we need to create a helper method in workflow-api for this retry logic that handles all of the cases correctly.

I totally agree with that. But this is kind of already done by the RetryStep:

https://github.com/jenkinsci/workflow-basic-steps-plugin/blob/686f0d7c0d9e235e958485a229552cc2da715626/src/main/java/org/jenkinsci/plugins/workflow/steps/RetryStepExecution.java

RetryStep works fine. It doesn't present this issue.

It would be nice if it could be reused here, but I am super inexperienced working with Jenkins plugins. I would appreciate if someone else can implement it, but please let me know if there is no intention in doing so.